src/index.ts

```diff
@@ -0,0 +1,87 @@
+export type StackErrorJSON = {
+  code: string
+  message: string
+  metadata: Record<string, unknown>
+  name: string
+  stack: Array<StackErrorJSON>
+}
+
+export class StackError extends Error {
+  static fromJSON(json: StackErrorJSON): StackError {
+    const error = new StackError(json.code, json.message)
+    error.errorStack = (json.stack ?? []).reduceRight((stack, e) => {
+      const err = StackError.fromJSON(e)
+      err.errorStack = stack
+      return [err, ...stack]
+    }, [] as Array<StackError>)
+    error.metadata = json.metadata ?? {}
+    error.name = json.name ?? 'StackError'
+    return error
+  }
+
+  code: string
+  errorStack: Array<StackError>
+  metadata: Record<string, unknown> = {}
+  name = 'StackError'
+
+  constructor(code: string, message: string, parentError?: StackError) {
+    super(message)
+    Object.setPrototypeOf(this, StackError.prototype)
+    if (Error.captureStackTrace) {
+      Error.captureStackTrace(this, StackError)
+    }
+
+    this.code = code
+    this.errorStack = parentError ? parentError.toErrorStack() : []
+  }
+
+  toErrorStack(): Array<StackError> {
+    return [this, ...this.errorStack]
+  }
+
+  toJSON(withStack = true): StackErrorJSON {
+    return {
+      code: this.code,
+      message: this.message,
+      metadata: this.metadata,
+      name: this.name,
+      stack: withStack ? this.errorStack.map(e => e.toJSON(false)) : []
+    }
+  }
+
+  toString(): string {
+    return `[${this.code}] ${this.message}`
+  }
+}
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+export function createNamespaceError(namespace: string, metadata: Record<string, unknown> = {}) {
+  return class NamespaceError extends StackError {
```

Comment:
I feel like each type of error should have its own class, rather than all being NamespaceErrors with a code field within, so that it becomes possible to write a try/catch block that catches only certain error types

Reply:
Yes that will be the case, here `NamespaceError` is just a local name for reference, but we can have something like:
```ts
const DocIDError = createNamespaceError('DCD')

class CeramicError extends createNamespaceError('CRC', { version: '0.2.3' }) {
  toAPIResponse(res: Response, statusCode = 500) {
    res.status(statusCode).json({ code: this.code, message: this.message })
  }
}

try {
  callFunction()
} catch (error) {
  if (error instanceof CeramicError) {
    ...
  } elseif (error instanceof DocIDError) {
    ...
  } elseif (error instanceof StackError) {
    ...
  }
}
```